From 34b0aa837f1a3d71a208106b3323bd479f11d532 Mon Sep 17 00:00:00 2001
From: Joseph Guo <qijian.guo@nxp.com>
Date: Wed, 6 Aug 2025 12:24:20 +0900
Subject: [PATCH 17/24] LFU-912-2: imx93_frdm: Add board version print

Add print_board_version() to print the board hardware version info in spl.

Signed-off-by: Joseph Guo <qijian.guo@nxp.com>
Signed-off-by: Tom Zheng <haidong.zheng@nxp.com>
Reviewed-by: Ye Li <ye.li@nxp.com>
---
 board/freescale/imx93_frdm/imx93_frdm.c | 47 +++++++++++++++++++++++++
 1 file changed, 47 insertions(+)

diff --git a/board/freescale/imx93_frdm/imx93_frdm.c b/board/freescale/imx93_frdm/imx93_frdm.c
index 573a13beca5..914573eb3f5 100644
--- a/board/freescale/imx93_frdm/imx93_frdm.c
+++ b/board/freescale/imx93_frdm/imx93_frdm.c
@@ -20,6 +20,7 @@
 #include <usb.h>
 #include <dwc3-uboot.h>
 #include <asm/gpio.h>
+#include <adc.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -263,8 +264,54 @@ static void board_gpio_init(void)
 	dm_gpio_set_value(&desc, 1);
 }
 
+static int print_board_version(void)
+{
+	int i, ret;
+	struct udevice *dev;
+	unsigned int rev[2];
+	unsigned int data[2];
+
+	ret = uclass_first_device_check(UCLASS_ADC, &dev);
+
+	if (dev) {
+		ret = adc_channel_single_shot(dev->name, 2, &data[0]);
+		if (ret) {
+			printf("BOARD: unknown\n");
+			return 0;
+		}
+		ret = adc_channel_single_shot(dev->name, 3, &data[1]);
+		if (ret) {
+			printf("BOARD: unknown\n");
+			return 0;
+		}
+
+		for (i = 0; i < 2; i++) {
+			if (data[i] < 500)
+				rev[i] = 0;
+			else if (data[i] < 700)
+				rev[i] = 1;
+			else if (data[i] < 1500)
+				rev[i] = 2;
+			else if (data[i] < 2300)
+				rev[i] = 3;
+			else if (data[i] < 3000)
+				rev[i] = 4;
+			else if (data[i] < 3600)
+				rev[i] = 5;
+			else
+				rev[i] = 6;
+		}
+		printf("BOARD: V%d.%d(ADC2:%d,ADC3:%d)\n", rev[0], rev[1], data[0], data[1]);
+	} else {
+		printf("BOARD: unknown\n");
+	}
+
+	return 0;
+}
+
 int board_init(void)
 {
+	print_board_version();
 #ifdef CONFIG_USB_TCPC
 	setup_typec();
 #endif
-- 
2.34.1

